File: claw-os-sdk/go/ai_test.go
package clawossdk

import (
	"slices"
	"strings"
	"testing"
)

const okChat = `{
  "verb": "ai.chat",
  "text": "hello there",
  "model": "m",
  "provider": "p",
  "usage": {"input_tokens": 3, "output_tokens": 5, "units": 8},
  "budget": {"period": "2026-05", "units_used": 8, "units_cap": 1000},
  "review": {"safety": "strict", "prompt_redacted": false},
  "tool_calls": [{"id": "c1", "name": "fs.read_text", "input": {"path": "/x"}}]
}`

func TestChatBuildsArgvAndParses(t *testing.T) {
	bin, argvOut := fakeCos(t, wireSuccess(okChat), 0)
	var res *AiResponse
	var err error
	withCos(t, bin, map[string]string{"COS_APP_ID": "notes"}, func() {
		res, err = Chat("summarise", ChatOptions{
			Origin:   "external-content",
			MaxUnits: 100,
			Tools:    []string{"fs.read_text"},
		})
	})
	if err != nil {
		t.Fatalf("Chat returned error: %v", err)
	}
	if res.Text != "hello there" {
		t.Fatalf("text = %q", res.Text)
	}
	if res.Usage.Units != 8 {
		t.Fatalf("units = %v", res.Usage.Units)
	}
	if res.Budget.UnitsCap != 1000 {
		t.Fatalf("units_cap = %v", res.Budget.UnitsCap)
	}
	if len(res.ToolCalls) != 1 || res.ToolCalls[0].Name != "fs.read_text" {
		t.Fatalf("tool_calls = %+v", res.ToolCalls)
	}

	argv := readArgv(t, argvOut)
	want := []string{"--wire=1", "ai", "chat", "--app", "notes", "--origin", "external-content"}
	for i, w := range want {
		if argv[i] != w {
			t.Fatalf("argv[%d] = %q, want %q (full %v)", i, argv[i], w, argv)
		}
	}
	if !slices.Contains(argv, "--prompt-file") || !slices.Contains(argv, "--max-units") || !slices.Contains(argv, "--tools") {
		t.Fatalf("missing expected flags: %v", argv)
	}
}

func TestChatEmptyPrompt(t *testing.T) {
	if _, err := Chat("   ", ChatOptions{AppID: "x"}); err == nil {
		t.Fatal("expected error for empty prompt")
	}
}

func TestChatRequiresAppID(t *testing.T) {
	bin, _ := fakeCos(t, wireSuccess(okChat), 0)
	var err error
	withCos(t, bin, map[string]string{"COS_APP_ID": ""}, func() {
		_, err = Chat("hi", ChatOptions{})
	})
	if _, ok := err.(*AiUnavailableError); !ok {
		t.Fatalf("expected AiUnavailableError, got %T %v", err, err)
	}
}

func TestChatErrorClassification(t *testing.T) {
	cases := []struct {
		body string
		want string
	}{
		{wireErrorEnvelope("safety words", "BUDGET_EXCEEDED"), "budget"},
		{wireErrorEnvelope("budget words", "SAFETY_VIOLATION"), "safety"},
		{wireErrorEnvelope("capability denied", "PERMISSION_DENIED"), "denied"},
	}
	for _, c := range cases {
		bin, _ := fakeCos(t, c.body, 1)
		var err error
		withCos(t, bin, map[string]string{"COS_APP_ID": "notes"}, func() {
			_, err = Chat("hi", ChatOptions{})
		})
		switch c.want {
		case "budget":
			if _, ok := err.(*AiBudgetExceededError); !ok {
				t.Fatalf("%s: want AiBudgetExceededError, got %T", c.body, err)
			}
		case "safety":
			if _, ok := err.(*AiSafetyViolationError); !ok {
				t.Fatalf("%s: want AiSafetyViolationError, got %T", c.body, err)
			}
		default:
			if _, ok := err.(*AiDeniedError); !ok {
				t.Fatalf("%s: want AiDeniedError, got %T", c.body, err)
			}
		}
	}
}

func TestChatRejectsMalformedToolCall(t *testing.T) {
	body := `{
	  "verb":"ai.chat","text":"hello","model":"m","provider":"p",
	  "usage":{"input_tokens":1,"output_tokens":1,"units":2},
	  "budget":{"period":"2026-08","units_used":2,"units_cap":100},
	  "review":{"safety":"strict","prompt_redacted":false},
	  "tool_calls":[{"id":"c1","input":{}}]
	}`
	bin, _ := fakeCos(t, wireSuccess(body), 0)
	var err error
	withCos(t, bin, map[string]string{"COS_APP_ID": "notes"}, func() {
		_, err = Chat("hi", ChatOptions{})
	})
	unavailable, ok := err.(*AiUnavailableError)
	if !ok || !strings.Contains(unavailable.Msg, "WIRE_REQUIRED") ||
		!strings.Contains(unavailable.Msg, "$.tool_calls[0].name") {
		t.Fatalf("expected strict decode error, got %T %v", err, err)
	}
}

func TestChatAcceptsMathematicalIntegersAndUnrestrictedToolInput(t *testing.T) {
	body := `{
	  "verb":"ai.chat","text":"hello","model":"m","provider":"p",
	  "usage":{"input_tokens":1.0,"output_tokens":1e0,"units":2.0},
	  "budget":{"period":"2026-08","units_used":2e0,"units_cap":100.0},
	  "review":{"safety":"strict","prompt_redacted":false},
	  "tool_calls":[{"id":"c1","name":"echo","input":["a",1]}]
	}`
	bin, _ := fakeCos(t, wireSuccess(body), 0)
	var response *AiResponse
	var err error
	withCos(t, bin, map[string]string{"COS_APP_ID": "notes"}, func() {
		response, err = Chat("hi", ChatOptions{})
	})
	if err != nil {
		t.Fatal(err)
	}
	if response.Usage.InputTokens != 1 || response.Usage.OutputTokens != 1 {
		t.Fatalf("usage = %+v", response.Usage)
	}
	if input, ok := response.ToolCalls[0].Input.([]any); !ok || len(input) != 2 {
		t.Fatalf("tool input = %#v", response.ToolCalls[0].Input)
	}
}

func TestChatRejectsScalarRootWithStableWireError(t *testing.T) {
	bin, _ := fakeCos(t, `null`, 0)
	var err error
	withCos(t, bin, map[string]string{"COS_APP_ID": "notes"}, func() {
		_, err = Chat("hi", ChatOptions{})
	})
	unavailable, ok := err.(*AiUnavailableError)
	if !ok || !strings.Contains(unavailable.Msg, "WIRE_ONE_OF") ||
		!strings.Contains(unavailable.Msg, " at $:") {
		t.Fatalf("expected stable root decode error, got %T %v", err, err)
	}
}

func TestBudgetAcceptsProducerShape(t *testing.T) {
	bin, _ := fakeCos(t, wireSuccess(`{"app":"notes","period":"2026-08","units_used":7}`), 0)
	var budget *AiBudget
	var err error
	withCos(t, bin, nil, func() {
		budget, err = Budget("notes")
	})
	if err != nil {
		t.Fatal(err)
	}
	if budget.Period != "2026-08" || budget.UnitsUsed != 7 || budget.UnitsCap != 0 {
		t.Fatalf("budget = %+v", budget)
	}
}

File: claw-os-sdk/node/src/ai.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";

import * as ai from "./ai";
import * as tools from "./tools";
import { WireDecimal, stringifyWireJson } from "./generated";
import {
  installFakeCos,
  wireError,
  wireSuccessJson,
  withCos,
} from "./testutil";

const OK_CHAT_DATA = JSON.stringify({
  verb: "ai.chat",
  text: "hello there",
  model: "m",
  provider: "p",
  usage: { input_tokens: 3, output_tokens: 5, units: 8 },
  budget: { period: "2026-05", units_used: 8, units_cap: 1000 },
  review: { safety: "strict", prompt_redacted: false },
  tool_calls: [{ id: "c1", name: "fs.read_text", input: { path: "/x" } }],
});
const OK_CHAT = wireSuccessJson(OK_CHAT_DATA);

test("chat builds the right argv and parses the envelope", () => {
  const fake = installFakeCos(OK_CHAT);
  const res = withCos(fake, { COS_APP_ID: "notes" }, () =>
    ai.chat("summarise", { origin: "external-content", maxUnits: 100, tools: ["fs.read_text"] }),
  );
  assert.equal(res.text, "hello there");
  assert.equal(res.usage.units, 8);
  assert.equal(res.budget.unitsCap, 1000);
  assert.equal(res.toolCalls.length, 1);
  assert.equal(res.toolCalls[0].name, "fs.read_text");

  const argv = readFileSync(fake.argvOut, "utf8").split("\n");
  assert.deepEqual(argv.slice(0, 7), [
    "--wire=1",
    "ai",
    "chat",
    "--app",
    "notes",
    "--origin",
    "external-content",
  ]);
  assert.ok(argv.includes("--prompt-file"));
  assert.ok(argv.includes("--max-units"));
  assert.ok(argv.includes("--tools"));
});

test("chat rejects an empty prompt", () => {
  assert.throws(() => ai.chat("  "), ai.AiError);
});

test("chat requires an app id", () => {
  const fake = installFakeCos(OK_CHAT);
  assert.throws(
    () => withCos(fake, { COS_APP_ID: undefined }, () => ai.chat("hi")),
    ai.AiError,
  );
});

test("budget-exceeded code maps to AiBudgetExceeded", () => {
  const fake = installFakeCos(wireError("opaque", "BUDGET_EXCEEDED"), 1);
  assert.throws(
    () => withCos(fake, { COS_APP_ID: "notes" }, () => ai.chat("hi")),
    ai.AiBudgetExceeded,
  );
});

test("safety code maps to AiSafetyViolation", () => {
  const fake = installFakeCos(wireError("opaque", "SAFETY_VIOLATION"), 1);
  assert.throws(
    () => withCos(fake, { COS_APP_ID: "notes" }, () => ai.chat("hi")),
    ai.AiSafetyViolation,
  );
});

test("generic error envelope maps to AiDenied", () => {
  const fake = installFakeCos(wireError("capability denied", "PERMISSION_DENIED"), 1);
  assert.throws(
    () => withCos(fake, { COS_APP_ID: "notes" }, () => ai.chat("hi")),
    ai.AiDenied,
  );
});

test("stable error code determines the typed error", () => {
  const budget = installFakeCos(wireError("safety words", "BUDGET_EXCEEDED"), 1);
  assert.throws(
    () => withCos(budget, { COS_APP_ID: "notes" }, () => ai.chat("hi")),
    ai.AiBudgetExceeded,
  );

  const safety = installFakeCos(wireError("budget words", "SAFETY_VIOLATION"), 1);
  assert.throws(
    () => withCos(safety, { COS_APP_ID: "notes" }, () => ai.chat("hi")),
    ai.AiSafetyViolation,
  );
});

test("malformed tool call rejects the entire response", () => {
  const payload = JSON.parse(OK_CHAT_DATA) as Record<string, unknown>;
  payload.tool_calls = [{ id: "c1", input: {} }];
  const fake = installFakeCos(wireSuccessJson(JSON.stringify(payload)));
  assert.throws(
    () => withCos(fake, { COS_APP_ID: "notes" }, () => ai.chat("hi")),
    (error: unknown) =>
      error instanceof ai.AiUnavailable &&
      error.message.includes("WIRE_REQUIRED") &&
      error.message.includes("$.tool_calls[0].name"),
  );
});

test("chat preserves unrestricted tool input and mathematical integers", () => {
  const payload = wireSuccessJson(OK_CHAT_DATA
    .replace('"input_tokens":3', '"input_tokens":1.0')
    .replace('"output_tokens":5', '"output_tokens":1e0')
    .replace('"units":8', '"units":18446744073709551615')
    .replace('"input":{"path":"/x"}', '"input":["a",1]'));
  const fake = installFakeCos(payload);
  const response = withCos(fake, { COS_APP_ID: "notes" }, () => ai.chat("hi"));
  assert.equal(response.usage.inputTokens, 1);
  assert.equal(response.usage.outputTokens, 1);
  assert.equal(response.usage.units, 18446744073709551615n);
  assert.deepEqual(response.toolCalls[0].input, ["a", 1]);
});

test("chat rejects scalar root with stable wire error", () => {
  const fake = installFakeCos("null");
  assert.throws(
    () => withCos(fake, { COS_APP_ID: "notes" }, () => ai.chat("hi")),
    (error: unknown) =>
      error instanceof ai.AiUnavailable &&
      error.message.includes("WIRE_ONE_OF") &&
      error.message.includes(" at $:"),
  );
});

test("budget accepts the producer budget-show shape", () => {
  const fake = installFakeCos(
    wireSuccessJson(JSON.stringify({ app: "notes", period: "2026-08", units_used: 7 })),
  );
  const result = withCos(fake, {}, () => ai.budget("notes"));
  assert.deepEqual(result, { period: "2026-08", unitsUsed: 7, unitsCap: 0 });
});

test("proposed lossless input round-trips directly through tools.call", () => {
  const lexeme = "0.12345678901234567890";
  const chatFake = installFakeCos(
    wireSuccessJson(OK_CHAT_DATA.replace('"input":{"path":"/x"}', `"input":${lexeme}`)),
  );
  const response = withCos(chatFake, { COS_APP_ID: "notes" }, () => ai.chat("hi"));
  assert.ok(response.toolCalls[0].input instanceof WireDecimal);
  assert.equal(stringifyWireJson(response.toolCalls[0].input), lexeme);

  const toolFake = installFakeCos(
    wireSuccessJson(
      JSON.stringify({ tool: "echo", app_id: "notes", status: "ok", result: null }),
    ),
  );
  withCos(toolFake, {}, () =>
    tools.call("echo", response.toolCalls[0].input, { appId: "notes" }),
  );
  const argv = readFileSync(toolFake.argvOut, "utf8").split("\n");
  assert.equal(argv[argv.indexOf("--args") + 1], lexeme);
});
